File: GFG/DP/27.Longest-Common-Substring.cpp
// Longest Common Substring

// Tabulation
class Solution{
    public:
    
    int longestCommonSubstr (string S1, string S2, int n, int m)
    {
        // your code here
        vector<vector<int>> dp(n+1, vector<int>(m+1,0));
        int ans = 0;
        for(int i=1; i<=n; i++) {
            for(int j=1; j<=m; j++) {
                if(S1[i-1] == S2[j-1]) {
                    dp[i][j] = 1 + dp[i-1][j-1];
                    ans = max(ans,dp[i][j]);
                } else {
                    dp[i][j] = 0;
                }
            }
        }
        return ans;
    }
};



// Space Optimization
class Solution{
    public:
    
    int longestCommonSubstr (string S1, string S2, int n, int m)
    {
        // your code here
        vector<int> dp(m+1,0), curr(m+1,0);
        int ans = 0;
        for(int i=1; i<=n; i++) {
            for(int j=1; j<=m; j++) {
                if(S1[i-1] == S2[j-1]) {
                    curr[j] = 1 + dp[j-1];
                    ans = max(ans,curr[j]);
                } else {
                    curr[j] = 0;
                }
            }
            dp = curr;
        }
        return ans;
    }
};

File: GFG/DP/27.Longest-Common-Substring.java
// Longest Common Substring

class Solution {
    public int longCommSubstr(String s1, String s2) {
        // code here
        int n1 = s1.length(), n2 = s2.length();
        int[][] dp = new int[n1][n2];
        
        int ans = 0;
        for (int i = 0; i < n1; i++) {
            for (int j = 0; j < n2; j++) {
                if (s1.charAt(i) == s2.charAt(j)) {
                    if (i == 0 || j == 0) dp[i][j] = 1;
                    else dp[i][j] = 1 + dp[i - 1][j - 1];
                }
                else {
                    dp[i][j] = 0;
                }
                ans = Math.max(ans, dp[i][j]);
            }
        }
        
        return ans;
    }
}
